### PR TITLE
Support configuration overrides

### PR DESCRIPTION
**Pull request by carterworks**
## Description

Support the configuration of datastream overrides in the Alloy Tags Extensions. Must be merged with https://github.com/adobe/alloy/pull/958.

Adds a section to the 

- Extension configuration
- Send event action
- Set consent action
- Redirect with identity action.

image

## Related Issue

https://jira.corp.adobe.com/browse/PDCL-9943

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.
- [x] I've updated the schema in extension.json or no changes are necessary.
- [x] My change requires a change to the documentation.

### CODE REVIEW

**Comment by carterworks** [comment]
The dropdowns prepopulated with the options from Blackbird will be implemented in [PDCL-10323](https://jira.corp.adobe.com/browse/PDCL-10323].

File: src/view/components/overrides.jsx [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  );
+};
+
+Overrides.propTypes = {
+  instanceFieldName: PropTypes.string,
+  largeHeader: PropTypes.bool,
+  showFields: PropTypes.arrayOf(PropTypes.oneOf(Object.values(FIELD_NAMES)))
+};
+
+export default Overrides;
```
Comment (jonsnyder) [inline src/view/components/overrides.jsx:273]: Somewhere we should indicate that the overrides have to be specified in the allow-list within the datastream configuration. I could see lots of people being surprised by this.

File: src/view/configuration/utils/copyPropertiesIfValueDifferentThanDefault.js [hunk 2]
```diff
@@ -9,20 +9,61 @@ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTA
...
+    const defaultValue = defaultsObj[key];
+
+    if (Array.isArray(fromValue) && Array.from(defaultValue)) {
+      if (fromValue.length !== defaultValue.length) {
+        toObj[key] = fromValue;
+        return;
+      }
+      // does not support arrays of objects
+      const sortedFromValue = fromValue.sort();
+      const sortedDefaultValue = defaultValue.sort();
```
Comment (jonsnyder) [inline src/view/configuration/utils/copyPropertiesIfValueDifferentThanDefault.js:40]: sort mutates the original array. (i.e. sortedFromValue and fromValue are the same). You'd have to do fromValue.slice(0).sort() if you didn't want it to mutate.

**Review by jonsnyder — APPROVED** [review]

File: src/view/configuration/utils/copyPropertiesIfValueDifferentThanDefault.js [hunk 3]
```diff
@@ -9,20 +9,61 @@ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTA
...
  * @param {Object} options.toObj Object to which properties should be copied.
  * @param {Object} options.fromObj Object from which properties should be copied.
  * @param {Object} options.defaultsObj Default values for each property.
- * @param {Array} options.keys The keys of the properties that should be attempted
- * to be copied.
+ * @param {Array} options.keys The top-level keys of the properties that should be attempted
+ * to be copied. Does not apply to nested keys
  */
-export default ({ toObj, fromObj, defaultsObj, keys }) => {
+export default function copyPropertiesIfValueDifferentThanDefault({
```
[the remark this answers is not in the document]
Comment (carterworks) [reply src/view/configuration/utils/copyPropertiesIfValueDifferentThanDefault.js:23]: In order to be able to call it recursively for nested objects (see line 56). There are ways of writing anonymous functions that can be recursive (y-combinators, apparently), but a named function is more simple.

**Review by ninaceban — APPROVED** [review]
Looks good to me 👍

**Pull request merged**